### ISSUE TITLE
week9 vgg net 구현 과제 질문입니다. 

### ISSUE THREAD

**Issue by hwanginsun** [public-ai/ALAI_Homework#73]
vgg net에 cifar10 dataset을 적용하는 과제 진행중입니다.

<br>

1. 현재 모델은 하기와 같이 구성해서 Model A에 대해서는 loss = 1.2375, **acc = 69.57%** 수준입니다.  

| Layer # |	Kernel  | Stride | Dilation	| Padding	| Input Size |Output Size | Receptive Field  |
|:-------:|:-------:|:------:|:------------:|:---------:|:---------:|:----------:|:---------:|
|    1	  |    3	| 1	     |      1	    |    2	    |  32    	| 32	     |     3  |
|    2	  |    3	| 1	     |      1	    |    2	    |  32    	| 32	     |     5  |
|    3	  |    3	| 1	     |      1	    |    2	    |  32    	| 32	     |     7  |
|    4	  |    3	| 1	     |      1	    |    2	    |  32    	| 32	     |     9  |
|    5	  |    2	| 2	     |      1	    |    0	    |  32    	| 16	     |     10  |
|    6	  |    3	| 3	     |      1	    |    1	    |  16    	| 6	     |     14  |
|    7	  |    2	| 2	     |      1	    |    0	    |  6	| 3 	     |     20  |

<br>

2. Model A에 대한 Train결과를 loss가 최저점일 때만 저장을 했습니다
```python3
if loss_ < minimum_loss 
   minimum_loss = loss_
  saver.save(sess, save_path = save_dir)
````
```python3
from google.colab import drive
drive.mount('/content/gdrive')

!mkdir gdrive/My\ Drive/vgg
!mv ./model/vgg* gdrive/My\ Drive/vgg
!mv ./model/checkpoint gdrive/My\ Drive/vgg
```

<br>

3. 최저점에서만 저장한 이유는 Model A의 결과에서  validation set에서 loss가 epoch이 20부근 이상부터 다시 올라가는 현상이 있었기 때문입니다. (model이 train set에 overfitting되가고 있었습니다)
![image](https://user-images.githubusercontent.com/30209848/59648457-20198a00-91ba-11e9-917e-1f585898510a.png)

<br>

4. Model B에서 train된 Model A의 결과를 하기와 같은 방식으로 loading했습니다
cp를 이용하여 값들을 numpy array로 불러와서 저장했습니다.

```python3
from google.colab import drive
drive.mount('/content/gdrive')

!mkdir ./model
!cp gdrive/My\ Drive/vgg/* model/     # from, to 임
```
```python3
from tensorflow.python.training import checkpoint_utils as cp

# 값 불러오기
conv1_w1  = cp.load_variable('./model/vgg_net_model_a', 'conv1/kernel1')   
conv1_b1  = cp.load_variable('./model/vgg_net_model_a', 'conv1/bias1')
conv1_w2  = cp.load_variable('./model/vgg_net_model_a', 'conv1/kernel2')   
conv1_b2  = cp.load_variable('./model/vgg_net_model_a', 'conv1/bias2')

conv2_w1  = cp.load_variable('./model/vgg_net_model_a', 'conv2/kernel1')   
conv2_b1  = cp.load_variable('./model/vgg_net_model_a', 'conv2/bias1')
conv2_w2  = cp.load_variable('./model/vgg_net_model_a', 'conv2/kernel2')   
conv2_b2  = cp.load_variable('./model/vgg_net_model_a', 'conv2/bias2')

conv3_w2  = cp.load_variable('./model/vgg_net_model_a', 'conv3/kernel2')   
conv3_b2  = cp.load_variable('./model/vgg_net_model_a', 'conv3/bias2')

f1_w1    = cp.load_variable('./model/vgg_net_model_a', 'Reconstruct/fc1_weights1')   
f1_b1    = cp.load_variable('./model/vgg_net_model_a', 'Reconstruct/fc1_bias1')
f2_w1    = cp.load_variable('./model/vgg_net_model_a', 'Reconstruct/fc2_weights1')   
f2_b1    = cp.load_variable('./model/vgg_net_model_a', 'Reconstruct/fc2_bias1')
```

<br>

5.  Model B를 훈련하기 위해 전체 초기화를 했고(Model B에는 momentum optimizer를 써서 이부분에 대한 초기화가 필요해서 이렇게 했습니다)
```python3
    sess = tf.Session()
    sess.run(tf.global_variables_initializer()) 
```

<br>

6. **feed_dict에 불러온 값을 인자로 넣어주었습니다**
먼저 제대로 값이 들어가는지 확인하기 위해 validation set으로 accurary를 확인하기 위해 아래와 같이 구성했습니다.
```python3
feed_dict = { xs : x_validation, ys : y_validation_label, 
                 wd : 0, m : 0, is_train : False, 
                 conv1_kernel1 : conv1_w1, conv1_bias1 : conv1_b1, 
                 conv1_kernel2 : conv1_w2, conv1_bias2 : conv1_b2, 
                 conv2_kernel1 : conv2_w1, conv2_bias1 : conv2_b1, 
                 conv2_kernel2 : conv2_w2, conv2_bias2 : conv2_b2, 
                 conv3_kernel2 : conv3_w2, conv3_bias2 : conv3_b2,
                 fc1_weights1  : f1_w1,    fc1_bias1 : f1_b1,
                 fc2_weights1  : f2_w1,    fc2_bias1 : f2_b1}
    loss_, acc_, fc2_weights1_= sess.run([rmse, accuracy, fc2_weights1], feed_dict = feed_dict)

    print("Loading결과 확인 -> loss = {:.4f}, acc = {:.2f}%".format(loss_, acc_*100))
    print("weight를 제대로 가져오는지 확인 -> ", fc2_weights1_)
```

<br>

7. **제대로 인자가 들어갔는지** 확인하기 위하여 Model A와 B의 weight 값을 찍어보았습니다
결과는 동일한 것으로 확인했습니다. (학습한 weight가 잘 들어가 있음)
![image](https://user-images.githubusercontent.com/30209848/59648667-d67d6f00-91ba-11e9-97ac-e90704c38115.png)

<br>

드디어 질문입니다..
**두개의 모델을 동일하게 세팅해놓은 상태인데, validation set에 대한 accuracy가 다르게 나옵니다.**
+ Model A : loss = 1.1984, acc = 68.54% (loss가 최저점일때이고, 저장할 때 결과임)
+ Model B :  loss = 1.5269, acc = 16.36%

model A : https://github.com/public-ai/ALAI_Homework/blob/c3cf18ef277546bc8b351ec2f1bfe06ab8f157e4/homework/week_9/insun/Week09_vgg_net_modelA.ipynb

model B : https://github.com/public-ai/ALAI_Homework/blob/c3cf18ef277546bc8b351ec2f1bfe06ab8f157e4/homework/week_9/insun/Week09_vgg_net_modelB.ipynb

어디부분을 확인하면 될지 가이드 부탁드립니다
감사합니다

model code 입니다
```python3
graph = tf.Graph()
with graph.as_default() :
    xs = tf.placeholder(tf.float32, (None, 32, 32, 3), name='xs') # 32*32 = 1024
    ys = tf.placeholder(tf.int32, (None), name='ys')
    lr = tf.placeholder_with_default(0.001, (), name='lr')
    wd = tf.placeholder_with_default(0.9, (), name='wd')
    is_train = tf.placeholder_with_default(False, (), name='is_train')
    m = tf.placeholder_with_default(0.9, (), name='momentum')
    
    with tf.name_scope('conv1') :   ## filters is number of channels
        kernel_init   = tf.zeros([3, 3, 3, 32]) 
        bias_init     = tf.zeros([32])
        conv1_kernel1 = tf.Variable(kernel_init, name='kernel1')               
        conv1_bias1   = tf.Variable(bias_init, name='bias1')
        
        layer         = tf.nn.conv2d(xs, conv1_kernel1, strides=[1,1,1,1], padding='SAME')
        layer         = layer + conv1_bias1
        layer         = tf.nn.relu(layer)
        
        kernel_init   = tf.zeros([3, 3, 32, 32]) 
        bias_init     = tf.zeros([32])
        conv1_kernel2 = tf.Variable(kernel_init, name='kernel2')     
        conv1_bias2   = tf.Variable(bias_init, name='bias2')
        
        layer         = tf.nn.conv2d(layer, conv1_kernel2, strides=[1,1,1,1], padding='SAME') 
        layer         = layer + conv1_bias2
        layer         = tf.nn.relu(layer)
          
            
    with tf.name_scope('conv2') :  
        kernel_init   = tf.zeros([3, 3, 32, 64]) 
        bias_init     = tf.zeros([64])
        conv2_kernel1 = tf.Variable(kernel_init, name='kernel1')    
        conv2_bias1   = tf.Variable(bias_init, name='bias1')    
        
        layer         = tf.nn.conv2d(layer, conv2_kernel1, strides=[1,1,1,1], padding='SAME') 
        layer         = layer + conv2_bias1
        layer         = tf.nn.relu(layer)
    
        kernel_init   = tf.zeros([3, 3, 64, 64]) 
        bias_init     = tf.zeros([64])
        conv2_kernel2 = tf.Variable(kernel_init, name='kernel2')   
        conv2_bias2   = tf.Variable(bias_init, name='bias2')    
        
        layer         = tf.nn.conv2d(layer, conv2_kernel2, strides=[1,1,1,1], padding='SAME')    
        layer         = layer + conv2_bias2
        layer         = tf.nn.relu(layer)
    
        pool          = tf.layers.MaxPooling2D(pool_size=[2,2], strides=[2,2])(layer)        

    with tf.name_scope('conv3') :       
        kernel_init   = tf.zeros([3, 3, 64, 128]) 
        bias_init     = tf.zeros([128])
        conv3_kernel2 = tf.Variable(kernel_init, name='kernel2')           
        conv3_bias2   = tf.Variable(bias_init, name='bias2')    
        
        layer         = tf.nn.conv2d(pool, conv3_kernel2, strides=[1,3,3,1], padding='SAME') 
        layer         = layer + conv3_bias2
        layer         = tf.nn.relu(layer)
    
        pool          = tf.layers.MaxPooling2D(pool_size=[2,2], strides=[2,2])(layer)      

    with tf.name_scope('Reconstruct') : 
        flatten       = tf.layers.flatten(pool)
        in_           = flatten.shape[1] 
        
        fc1_weights1  = tf.Variable(tf.zeros(shape = [in_, 1024], dtype = tf.float32), name = "fc1_weights1")
        fc1_bias1     = tf.Variable(tf.zeros(shape = [1024], dtype = tf.float32), name = "fc1_bias1")
        fc1_logits    = tf.matmul(flatten, fc1_weights1) + fc1_bias1
        fc1_act       = tf.nn.relu(fc1_logits)
        fc1_dropout   = tf.layers.Dropout(0.5)(fc1_act, training = is_train)
        fc1_result    = tf.identity(fc1_dropout, "fc1_result")
        
        fc2_weights1  = tf.Variable(tf.zeros(shape = [1024, 1024], dtype = tf.float32), name = "fc2_weights1")
        fc2_bias1     = tf.Variable(tf.zeros(shape = [1024], dtype = tf.float32), name = "fc2_bias1")
        fc2_logits    = tf.matmul(fc1_result, fc2_weights1) + fc2_bias1
        fc2_act       = tf.nn.relu(fc2_logits)
        fc2_dropout   = tf.layers.Dropout(0.5)(fc2_act, training = is_train)
        fc2_result    = tf.identity(fc2_dropout, "fc2_result")
        
        y_pred        = tf.layers.Dense(10, activation=None, name='y_pred')(fc2_result) 
    # Layer #	Kernel Stride	Dilation	Padding	Input Size	Output Size	Receptive Field
    #    1	      3	       1	   1	       2	      32    	 32	          3
    #    2	      3	       1	   1	       2	      32    	 32	          5
    #    3	      3	       1	   1	       2	      32    	 32	          7
    #    4	      3	       1	   1	       2	      32    	 32	          9
    #    5	      2	       2	   1	       0	      32    	 16	          10
    #    6	      3	       3	   1	       1	      16    	 6	          14
    #    7	      2	       2	   1	       0	      6	         3 	          20       
    
    with tf.name_scope('Loss') :
        sce_loss      = tf.reduce_mean(tf.losses.sparse_softmax_cross_entropy(labels=ys, logits=y_pred))
        l2_loss       = tf.add_n([tf.nn.l2_loss(var) for var in tf.global_variables()])
        loss          = sce_loss + wd * l2_loss
    loss = tf.identity(loss, name='loss')
        
    with tf.name_scope('metric') :
        rmse = tf.sqrt(loss)
    
    with tf.name_scope('accuracy') :
        pred        = tf.cast(tf.arg_max(y_pred, 1), tf.int32)
        correct     = tf.equal(pred, ys)
        accuracy    = tf.reduce_mean(tf.cast(correct, tf.float32))
        
        # add tensor to tensorboard
        acc_tb      = tf.summary.scalar(name='accuracy', tensor=accuracy)

    with tf.name_scope('train') :
        global_step = tf.train.get_or_create_global_step()
        #train_op   = tf.train.AdamOptimizer(lr).minimize(loss, global_step = global_step)
        
        optimizer   = tf.train.MomentumOptimizer(lr, momentum = m, use_nesterov = False)
        train_op    = optimizer.minimize(loss, global_step = global_step)
```

```python3
with graph.as_default() :
    
    sess = tf.Session()
    sess.run(tf.global_variables_initializer()) 

    # Training
    batch_size = 1000
    n_epoch = 0 # 30
    n_step = int(len(x_train) // batch_size)  # //은 몫이다.
    learing_rate = 0.001
    weight_decay = 0.0005
    
    # instance 생성
    train_generator = cifar_generator(x_train, y_train_label, batch_size)
    
    train_loss = []
    valid_loss = []
    valid_acc = []
    cnt = 0
    minimum_loss = 1.1
    momentum = 0.9
    
    # pre training된 weight, bais 넣기 ## Numpy array로 직접 넣기
    feed_dict = { xs : x_validation, ys : y_validation_label, 
                 wd : 0, m : 0, is_train : False, 
                 conv1_kernel1 : conv1_w1, conv1_bias1 : conv1_b1, 
                 conv1_kernel2 : conv1_w2, conv1_bias2 : conv1_b2, 
                 conv2_kernel1 : conv2_w1, conv2_bias1 : conv2_b1, 
                 conv2_kernel2 : conv2_w2, conv2_bias2 : conv2_b2, 
                 conv3_kernel2 : conv3_w2, conv3_bias2 : conv3_b2,
                 fc1_weights1  : f1_w1,    fc1_bias1 : f1_b1,
                 fc2_weights1  : f2_w1,    fc2_bias1 : f2_b1}
    loss_, acc_, fc2_weights1_= sess.run([rmse, accuracy, fc2_weights1], feed_dict = feed_dict)
    print("Loading결과 확인 -> loss = {:.4f}, acc = {:.2f}%".format(loss_, acc_*100))
    print("weight를 제대로 가져오는지 확인 -> ", fc2_weights1_)
```

**Comment by pai-plznw4me** [comment]
1. weights feed_dict 으로 넣지 말아주세요! 다른 방법으로 넣어주셔야 합니다. 
tf.Variable 의 초기값으로 Numpy 을 넣어줄수 있다는 점을 상기해보시면 답을 쉽게 찾으실거라 생각합니다. 

2. accuracy 가 너무 낮습니다. 85% 도달 못하실거에요 . pretraining 코드에  augmentataion 넣어주세요.

3. reconstruct model 에서 이렇게 작성하신 이유가 뭘까요? 
즉 질문을 정리하사면 왜 2개의 conv layer 을 작성했는지 궁굼합니다. 
```
 with tf.name_scope('conv1') :   ## filters is number of channels
        kernel_init   = tf.zeros([3, 3, 3, 32]) 
        bias_init     = tf.zeros([32])
        conv1_kernel1 = tf.Variable(kernel_init, name='kernel1')               
        conv1_bias1   = tf.Variable(bias_init, name='bias1')
        
        layer         = tf.nn.conv2d(xs, conv1_kernel1, strides=[1,1,1,1], padding='SAME')
        layer         = layer + conv1_bias1
        layer         = tf.nn.relu(layer)
        
        kernel_init   = tf.zeros([3, 3, 32, 32]) 
        bias_init     = tf.zeros([32])
        conv1_kernel2 = tf.Variable(kernel_init, name='kernel2')     
        conv1_bias2   = tf.Variable(bias_init, name='bias2')
        
        layer         = tf.nn.conv2d(layer, conv1_kernel2, strides=[1,1,1,1], padding='SAME') 
        layer         = layer + conv1_bias2
        layer         = tf.nn.relu(layer)
          `
```
그리고 해당 [모델 B](https://github.com/publicai/ALAI_Homework/blob/c3cf18ef277546bc8b351ec2f1bfe06ab8f157e4/homework/week_9/insun/Week09_vgg_net_modelB.ipynb) 코드 읽기가 너무 어렵습니다. 선생님의 논리가 보이질 않아요. 
주석 달아주시고. 필요없는 코드 다 지워주세요.

**Comment by hwanginsun** [comment]
1. Model B에 대해서 refactoring진행하고 있습니다.
2. tf.Variable에 numpy에 값을 넣을 수 있도록, model안에서 weights를 불러와서 넣는 방식으로 진행했습니다.
3. Model A에 대한 accuracy는 ModelA와 B에 대해서 코드 정리 이후에 진행하겠습니다. 
4. conv-layer가 2개인 이유는 이것저것 try해보다 진행했습니다. vgg net과 동일하게 진행하겠습니다.

```python3
graph = tf.Graph()
with graph.as_default() :
    xs = tf.placeholder(tf.float32, (None, 32, 32, 3), name='xs') 
    ys = tf.placeholder(tf.int32, (None), name='ys')
    lr = tf.placeholder_with_default(0.001, (), name='lr')
    wd = tf.placeholder_with_default(0.9, (), name='wd')
    is_train = tf.placeholder_with_default(False, (), name='is_train')
    m = tf.placeholder_with_default(0.9, (), name='momentum')
    
    conv1_w1  = cp.load_variable('./model/vgg_net_model_a', 'conv1/kernel1')   
    conv1_b1  = cp.load_variable('./model/vgg_net_model_a', 'conv1/bias1')

    with tf.name_scope('conv1') :  
        kernel = tf.Variable(conv1_w1, name='kernel1')               
        bias   = tf.Variable(conv1_b1, name='bias1')        
        layer  = tf.nn.conv2d(xs, kernel, strides=[1,1,1,1], padding='SAME')
        layer  = layer + bias
        layer  = tf.nn.relu(layer)
```

**Comment by pai-plznw4me** [comment]
*conv-layer가 2개인 이유는 이것저것 try해보다 진행했습니다. vgg net과 동일하게 진행하겠습니다.*
선생님의 의도하신바가 있다면 진행하셔도 상관없습니다. 코드가 잘못된 코드인줄 알았습니다.

**Comment by hwanginsun** [comment]
context manage의 naming을 변경해서 명확하게 코드리뷰할 수 있도록 하겠습니다.
감사합니다

**conv1 -> VGGBlock1**
